src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Resolution.cs

```diff
@@ -925,15 +925,19 @@ private static void AddModule(PEModule module, int referenceIndex, ResolvedRefer
 
             // definition with the lowest version higher than reference version, unless exact version found
             int minHigherVersionDefinition = -1;
+            Version? minHigherVersionDefinitionVersion = null;
             int maxLowerVersionDefinition = -1;
+            Version? maxLowerVersionDefinitionVersion = null;
 
-            // Skip assembly being built for now; it will be considered at the very end:
-            bool resolveAgainstAssemblyBeingBuilt = definitionStartIndex == 0;
-            definitionStartIndex = Math.Max(definitionStartIndex, 1);
-
-            for (int i = definitionStartIndex; i < definitions.Length; i++)
+            foreach ((AssemblyData definitionData, int definitionIndex) in definitions[reference.Name])
             {
-                AssemblyIdentity definition = definitions[i].Identity;
+                // Skip assembly being built for now; it will be considered at the very end
+                if (definitionIndex == 0)
+                {
+                    continue;
+                }
+
+                AssemblyIdentity definition = definitionData.Identity;
 
                 if (!AssemblyIdentityComparer.SimpleNameComparer.Equals(reference.Name, definition.Name))
```

Comment:
>if (!AssemblyIdentityComparer.SimpleNameComparer.Equals(reference.Name, definition.Name) [](http://example.com/codeflow?start=16&length=88)

Is this still necessary? If not, consider removing or converting to an assert.

Reply:
This was the fast-path I added to try and move a common comparison check up sooner. I think we can remove it, and not have an assert. It will be verified by `assemblyIdentityComparer` anyway.

Reply:
OK, will remove completely